### DISCUSSION TITLE
Bounding Box

### DISCUSSION THREAD

**Discussion by daniel-wong-dfinity-org** [mgeisler/textwrap#531] [Ideas]
I would like a function like this:

```rust
fn bound(s: &str, max_width: u64, max_height: u64) -> String {
    ...
}

fn bound_debug(object: &impl Debug, max_width: u64, max_height: u64) -> String {
    bound(format!("{:#?}", object), max_width, max_height);
}
```

If s is small, a copy gets returned. But if it has too many lines, the middle ones get replaced with "...". Similarly, if a line within s is too long, the right portion of it gets replaced with "...". Would you accept a pull request for this?

# Use Case

Something goes wrong, so I want to log it. The log message should contain breadcrumbs. Often, the breadcrumb information is one or more object(s). But the object(s) could be big, and I do not want overwhelmingly large log messages. So, I would like to do this:

```rust
log!("Invalid:\n{}", bound_debug(&breadcrumb, 80, 10));
```

**Comment by mgeisler** [comment]
Hi @daniel-wong-dfinity-org, thanks for writing!

> If s is small, a copy gets returned. But if it has too many lines, the middle ones get replaced with "...". Similarly, if a line within s is too long, the right portion of it gets replaced with "...". Would you accept a pull request for this?

If I understand you correctly, then you're looking for something similar to this:

```rust
fn bound(s: &str, max_width: usize, max_height: usize) -> String {
    let mut lines = Vec::new();

    let options = textwrap::Options::new(max_width.saturating_sub(3))
        .wrap_algorithm(textwrap::wrap_algorithms::WrapAlgorithm::FirstFit);

    for line in s.lines() {
        if textwrap::core::display_width(line) <= max_width {
            lines.push(String::from(line));
        } else {
            lines.push(format!("{}...", textwrap::wrap(line, &options)[0]));
        }
    }

    if lines.len() > max_height {
        format!(
            "{}\n...\n{}",
            &lines[..max_height / 2].join("\n"),
            &lines[lines.len() - max_height / 2..].join("\n")
        )
    } else {
        lines.join("\n")
    }
}
```

This is quite brute-force: it wraps the entire long line just to extract the first of the wrapped lines.

I don't think this fits super well in Textwrap: it's not so much about wrapping as it is about truncating.

I would prefer if this could go into a different tiny crate for "error formatting" or similar. That way Textwrap can stay focused on being a toolbox with primitives that others can use for building more expressive UIs.

**Reply by daniel-wong-dfinity-org** [reply]
> Hi @daniel-wong-dfinity-org, thanks for writing!

😁 

> I would prefer if this could go into a different tiny crate

Ok.

> it's not so much about wrapping as it is about truncating.

Correct.

I guess I figured it might fit, because one reason ppl reflow their text is so that it won't be too wide. What I'm suggesting is also for when text might be too wide (and/or too high).

**Reply by mgeisler** [reply]
> I guess I figured it might fit, because one reason ppl reflow their text is so that it won't be too wide. What I'm suggesting is also for when text might be too wide (and/or too high).

I see what you mean — I'm just afraid that there will be too many choices here that it becomes hard to make a reusable solution which is helpful for lots of people.

Even with the code above, there will likely be some who don't want the text to be trimmed with `...` in the middle — some might find it more natural to keep the lines that fit and then add a `...` as the last line if there were too many.

So I think Textwrap can provide some of the tools you need for this, but the details of how you combine these tools is likely different from application to application.
